core/execution/market_snapshot.go

```diff
@@ -161,6 +161,10 @@ func NewMarketFromSnapshot(
 	market.tradableInstrument.Instrument.Product.NotifyOnSettlementPrice(market.settlementPrice)
 	liqEngine.SetGetStaticPricesFunc(market.getBestStaticPricesDecimal)
 
+	if mkt.State == types.MarketStateTradingTerminated {
+		market.tradableInstrument.Instrument.Product.UnsubscribeTradingTerminated(ctx)
+	}
```

Comment:
Why not d othe check at line 160, and just not register in the first time?

Reply:
160 is the callback from the instrument into the market. The subscription to the oracles happens deep in the heart of `markets.NewTradableInstrument(ctx, log, mkt.TradableInstrument, oracleEngine)`: https://github.com/vegaprotocol/vega/blob/develop/core/products/future.go#L260

Reply:
right makes sense. just feels a bit weird that we register it just to unregister soon after.

Reply:
I hear you. I'll do a whole `NewTradableInstrumentFromSnapshot()` type thing.

Reply:
ah sounds good.